apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/ShardDownloadPresenceObserver.java

```diff
@@ -137,11 +137,14 @@ public void onNoShardPresent(@Observes @ShardPresentEvent(ShardPresentEventType.
                     dataSource.begin();
                 }
                 try (Connection con = dataSource.getConnection()) {
+                    // create first genesis block, but do not save it to db here
                     Block genesisBlock = genesisImporter.newGenesisBlock();
-                    addBlock(dataSource, genesisBlock);
                     long initialBlockId = genesisBlock.getId();
                     log.debug("Generated Genesis block with Id = {}", initialBlockId);
+                    // import other genesis data
                     genesisImporter.importGenesisJson(false);
+                    // first genesis block should be saved only after all genesis data has been imported before
+                    addBlock(dataSource, genesisBlock); // save first genesis block here
```

Comment:
When genesis data was imported successfully but exception occurred during adding genesis block node will shutdown and after next startup will duplicate genesis data because all checks pass successfully.

Reply:
I updated code, data clean up is always executed in GenesisImporter. If genesis block is in DB and didn't fail on save we will not get into GenesisImport.
I tested code run for failed genesis block save.